File: advent-of-code/2025/day-6-trash-compactor/normal/n-d06-12.py
import sys

def solve(data):
    lines = data.strip().split('\n')
    max_len = max(len(line) for line in lines)
    lines = [line.ljust(max_len) for line in lines]
    
    cols = [''.join(lines[r][c] for r in range(len(lines))) for c in range(max_len)]
    
    problems = []
    i = 0
    while i < len(cols):
        if cols[i]. strip() == '':
            i += 1
            continue
        j = i
        while j < len(cols) and cols[j].strip() != '':
            j += 1
        prob_cols = cols[i:j]
        op = prob_cols[0][-1]
        nums = []
        for col in prob_cols:
            num_str = col[:-1]. strip()
            if num_str: 
                nums.append(int(num_str))
        problems.append((op, nums))
        i = j
    
    p1 = 0
    for op, nums in problems:
        if op == '+':
            p1 += sum(nums)
        else: 
            r = 1
            for n in nums:
                r *= n
            p1 += r
    cols_rev = cols[::-1]
    problems2 = []
    i = 0
    while i < len(cols_rev):
        if cols_rev[i].strip() == '':
            i += 1
            continue
        j = i
        while j < len(cols_rev) and cols_rev[j].strip() != '':
            j += 1
        prob_cols = cols_rev[i:j]
        op = prob_cols[0][-1]
        nums = []
        for col in prob_cols: 
            digits = col[:-1].replace(' ', '')
            if digits: 
                nums.append(int(digits))
        problems2.append((op, nums))
        i = j
    
    p2 = 0
    for op, nums in problems2:
        if op == '+':
            p2 += sum(nums)
        else:
            r = 1
            for n in nums: 
                r *= n
            p2 += r
    
    return p1, p2

if __name__ == "__main__": 
    p1, p2 = solve(sys.stdin.read())
    print(p1)
    print(p2)

File: advent-of-code/2025/day-6-trash-compactor/normal/n-d06-12.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    vector<string> lines;
    string line;
    size_t maxlen = 0;
    while (getline(cin, line)) {
        lines.push_back(line);
        maxlen = max(maxlen, line.size());
    }
    for (auto& l : lines) l.resize(maxlen, ' ');
    
    vector<string> cols(maxlen);
    for (size_t c = 0; c < maxlen; c++)
        for (size_t r = 0; r < lines.size(); r++)
            cols[c] += lines[r][c];
    
    auto solve = [&](vector<string>& cs) {
        long long total = 0;
        size_t i = 0;
        while (i < cs.size()) {
            while (i < cs.size() && cs[i]. find_first_not_of(' ') == string::npos) i++;
            if (i >= cs.size()) break;
            size_t j = i;
            while (j < cs.size() && cs[j].find_first_not_of(' ') != string::npos) j++;
            char op = cs[i]. back();
            vector<long long> nums;
            for (size_t k = i; k < j; k++) {
                string s = cs[k]. substr(0, cs[k]. size() - 1);
                s.erase(remove(s. begin(), s.end(), ' '), s.end());
                if (! s.empty()) nums.push_back(stoll(s));
            }
            if (op == '+') {
                long long sum = 0;
                for (auto n : nums) sum += n;
                total += sum;
            } else {
                long long prod = 1;
                for (auto n : nums) prod *= n;
                total += prod;
            }
            i = j;
        }
        return total;
    };
    
    cout << solve(cols) << "\n";
    reverse(cols.begin(), cols.end());
    cout << solve(cols) << "\n";
}
